README.md
# Aramok's Clearcase

## 🐲  Aramok's Clearcase  
> - Under Development....


### 📜 Questions, Feedback, Contributing
>- Feel free to send message for suggestions



CHANGELOG.md
# Change Log

All notable changes to the "aramoks-black" extension will be documented in this file.

## [Unreleased]

- Minor bug fix (Status bar info stuck)
- Username information Statusbar file status info added
- History tree feature added
- Prior time latest history info status bar text added
- Symbolic File Support Added
- Make element context visibility bug fixed
- Information and Error notifications are now optional
- Statusbar text visibility bug solved
- Fix bug if space exist in comment input
- Version status text bug fixed
- Private files are not distinguishable
- Shortcut for quick describe from status bar added
- File version info fixed
- Comment option dialog box added
- Settings Added
- Bug fix! @check issues
- Application icon added
- cleartool , checkout , checkin , undo checkout , make element ... added
- cleartool path check
- Context menu added
- Initial Commit


src/common.ts
'use strict';
import * as vscode from 'vscode';
import * as fs from 'fs';
import { LogCat } from './logcat';

export enum Glyph {
	PlusSign = '\u002B',
	ReturningArrow = '\u2937',
	DownwardsArrow = '\u2193',
	CheckMark = '\u2713',
	Bullet = '\u2022',
	TripleLine = '\u2261',
	EarthGround = '\u23DA'
}

export function fetchRealLocation(path: fs.PathLike): Promise<string> {
	return new Promise((resolve, reject) => {
		let realFile: string;
		fs.lstat(path, function (err, stats) {
			if (err) {
				reject(err);
			}
			if (stats.isSymbolicLink()) {
				fs.readlink(path, (err: NodeJS.ErrnoException | null, linkstring: string) => {
					LogCat.getInstance().log('Symbolic File: ' + path);
					resolve('"' + linkstring + '"');
				});
			} else {
				LogCat.getInstance().log('File: ' + path);
				resolve('"' + path.toString() + '"');
			}
		});

	});
}

export function datePriorToNow(date: Date | RegExpMatchArray | null): string {
	if (date) {
		if (date instanceof Date) {
			let dateTime: number = ((new Date()).valueOf() - date.valueOf());
			var msPerMinute = 60 * 1000;
			var msPerHour = msPerMinute * 60;
			var msPerDay = msPerHour * 24;
			var msPerMonth = msPerDay * 30;
			var msPerYear = msPerDay * 365;
			if (dateTime < msPerMinute) {
				return ' ' + Math.round(dateTime / 1000) + ' seconds ago';
			} else if (dateTime < msPerHour) {
				return ' ' + Math.round(dateTime / msPerMinute) + ' minutes ago';
			} else if (dateTime < msPerDay) {
				return ' ' + Math.round(dateTime / msPerHour) + ' hours ago';
			} else if (dateTime < msPerMonth) {
				return ' ~' + Math.round(dateTime / msPerDay) + ' days ago';
			} else if (dateTime < msPerYear) {
				return ' ~' + Math.round(dateTime / msPerMonth) + ' months ago';
			} else {
				return ' ~' + Math.round(dateTime / msPerYear) + ' years ago';
			}
		} else {
			return datePriorToNow(new Date(date[0].toString()));
		}
	} else {
		return ' - ';
	}
}

src/userinterface.ts
'use strict';
import * as vscode from 'vscode';
import { LogCat } from './logcat';
import * as config from './configuration';

export enum NotificationType {
	Information,
	Warning,
	Error
}

export enum FileState {
	Locked,
	CheckedOut,
	Private,
	Unknown
}

export enum StatusBarItemType {
	ViewStatus,
	FileStatus,
	FileState,
	None
}

export let viewStatus: vscode.StatusBarItem;
export let fileStatus: vscode.StatusBarItem;
export let fileState: vscode.StatusBarItem;

export function iniStatusBar() {
	viewStatus = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 2);
	viewStatus.tooltip = "Version info of current element";
	viewStatus.text = `$(repo-sync~spin) Activating Cleartool...`;

	fileStatus = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 1);
	fileStatus.command = "cleartool.showHistory";
	fileStatus.tooltip = "Click for Full History";

	fileState = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 0);
	fileState.command = "extension.describe";
	fileState.tooltip = "Refresh File Status";

	viewStatus.show();
	fileStatus.show();
	fileState.show();
}

export function hideStatusBar() {
	fileState.hide();
	fileStatus.hide();
	viewStatus.hide();
}

interface Progression {
	location: vscode.ProgressLocation.Notification;
	title: String;
	cancellable: boolean;
}


export function progressStatusBar(type: StatusBarItemType | undefined, progressText: string,
	callback?: (progress: vscode.Progress<{ message?: string | undefined; increment?: number | undefined; }>, finish: () => void) => void,
	) {

	let restore: string;
	config.checkConfigForStatusBar(fileState, 'statusBarShowFileState');
	config.checkConfigForStatusBar(fileStatus, 'statusBarShowFileInfo');
	config.checkConfigForStatusBar(viewStatus, 'statusBarShowViewStatus');

	switch (type) {
		case StatusBarItemType.ViewStatus:
			restore = viewStatus.text;
			viewStatus.text = `$(repo-sync~spin) ` + progressText + '...';
			break;
		case StatusBarItemType.FileStatus:
			restore = fileStatus.text;
			fileStatus.text = `$(repo-sync~spin) ` + progressText + '...';
			break;
		case StatusBarItemType.FileState:
			restore = fileState.text;
			fileState.text = `$(repo-sync~spin) ` + progressText + '...';
			break;
		default:
			break;
	}

	//use restore with promose !keep your promises

	let progressOptions: vscode.ProgressOptions = {
		location: vscode.ProgressLocation.Notification,
		title: progressText,
		cancellable: false
	};

	if (callback) {
		vscode.window.withProgress(progressOptions, (progress, token) => {
			
			var p = new Promise(resolve => {
				callback(progress , resolve);
			});
			return p;
		});
	}
}


/*
if(finish){
	finish(p = new Promise<void>);
	return p;
}else{
	p = new Promise(resolve => {
		setTimeout(() => {
			resolve();
		}, 15000);
	});
	return p;
}
*/


export function showMessage(message: String, type: NotificationType) {
	switch (type) {
		case NotificationType.Information:
			if (vscode.workspace.getConfiguration("cleartool").get("showInformationMessages")) {
				vscode.window.showInformationMessage(`${message}`, 'Don\'t show informations again').then(selection => {
					if (selection) {
						if (selection === "Don't show informations again") {
							vscode.workspace.getConfiguration("cleartool").update('showInformationMessages', false, vscode.ConfigurationTarget.Global);
							showOpenSettings("Information");
						}
					}
				});
			}
			break;
		case NotificationType.Error:
			if (vscode.workspace.getConfiguration("cleartool").get("showErrorMessages")) {
				vscode.window.showErrorMessage(`Error : ${message}`, 'Don\'t show informations again').then(selection => {
					if (selection) {
						if (selection === 'Don\'t show errors again') {
							vscode.workspace.getConfiguration("cleartool").update('showErrorMessages', false, vscode.ConfigurationTarget.Global);
							showOpenSettings("Error");
						}
					}
				});
			}
			break;
		default: case NotificationType.Warning:
			break;
	}
	LogCat.getInstance().log(`Result: ${message}`);
}

export function showOpenSettings(type: String) {
	vscode.window.showInformationMessage("" + type + " notification feedbacks for cleartool are not visible anymore. In case you change your mind , you can find parameters to change behaviour of extension in the settings section.", 'Understood', 'Go to Settings').then(selection => {
		if (selection) {
			if (selection === 'Go to Settings') {
				vscode.commands.executeCommand('workbench.action.openSettings', 'cleartool');
			}
		}
	});
}

export function setContextCriteria(state: FileState) {
	switch (state) {
		case FileState.Locked:
			vscode.commands.executeCommand('setContext', 'mkelem-criteria', false);
			vscode.commands.executeCommand('setContext', 'checkin-criteria', false);
			vscode.commands.executeCommand('setContext', 'checkout-criteria', true);
			break;
		case FileState.CheckedOut:
			vscode.commands.executeCommand('setContext', 'mkelem-criteria', false);
			vscode.commands.executeCommand('setContext', 'checkin-criteria', true);
			vscode.commands.executeCommand('setContext', 'checkout-criteria', false);
			break;
		case FileState.Private:
			vscode.commands.executeCommand('setContext', 'mkelem-criteria', true);
			vscode.commands.executeCommand('setContext', 'checkin-criteria', false);
			vscode.commands.executeCommand('setContext', 'checkout-criteria', false);
			break;
		case FileState.Unknown: default:
			vscode.commands.executeCommand('setContext', 'mkelem-criteria', false);
			vscode.commands.executeCommand('setContext', 'checkin-criteria', false);
			vscode.commands.executeCommand('setContext', 'checkout-criteria', false);
			break;
	}
}

export function showCommentDialog(callback: (val: string) => void) {
	if (vscode.workspace.getConfiguration("cleartool").get("actionCommentOptionDialog")) {
		vscode.window.showInputBox({
			prompt: "Type Comment to Check-Out File ",
			placeHolder: "Cannot be empty (see extension settings)"
		}).then(value => {
			if (value) {
				callback('-c ' + '"' + value + '"');
				return;
			} else {
				return;
			}
		});
	} else {
		callback("-nc");
		return;
	}
}

export async function showUncoQuickPick(callback: (val: string) => void) {
	let i = 0;
	const result = await vscode.window.showQuickPick(['Yes', 'No'], {
		placeHolder: 'Keep copy of your changes?'
	});

	switch (result) {
		case 'Yes':
			callback('-keep');
			break;
		default: case 'No':
			callback('-rm');
			break;
	}
}


export async function showInputBox() {
	const result = await vscode.window.showInputBox({
		value: 'abcdef',
		valueSelection: [2, 4],
		placeHolder: 'For example: fedcba. But not: 123',
		validateInput: text => {
			vscode.window.showInformationMessage(`Validating: ${text}`);
			return text === '123' ? 'Not 123!' : null;
		}
	});
	vscode.window.showInformationMessage(`Got: ${result}`);
}



src/historyProvider.ts
'use strict';
import { exec, spawn, ChildProcess, ExecException } from 'child_process';
import * as vscode from 'vscode';
import * as path from 'path';
import { Glyph } from './common';


export interface HistoryData {
	data: HistoryData[];
	version: string;
	username: string;
	operation: string;
	type: string;
	time: string;
	icon: string;
	comment?: string;
}

enum TreeItemType {
	Root,
	View,
	Element
}

export class HistoryProvider implements vscode.TreeDataProvider<Element>{
	historyData: HistoryData[];

	private _onDidChangeTreeData: vscode.EventEmitter<Element | undefined> = new vscode.EventEmitter<Element | undefined>();
	onDidChangeTreeData?: vscode.Event<Element | null | undefined> | undefined = this._onDidChangeTreeData.event;

	constructor() {
		this.historyData = [];
	}

	refresh(): void {
		this._onDidChangeTreeData.fire();
	}

	fetchHistory(historyData: HistoryData[]): void {
		this.historyData = historyData;
		this._onDidChangeTreeData.fire();
	}

	getTreeItem(element: Element): vscode.TreeItem | Thenable<vscode.TreeItem> {
		return element;
	}

	getChildren(element?: Element): Thenable<Element[]> {
		let deps: Element[] = [];
		if (element) {
			switch (element.collapsibleState) {
				case vscode.TreeItemCollapsibleState.Collapsed:
				case vscode.TreeItemCollapsibleState.Expanded:
					element.historyData.forEach(history => {
						if (history.data.length > 0) {
							deps.push(new Element(
								history.version, ' ', TreeItemType.View, history.data, vscode.TreeItemCollapsibleState.Expanded));
						} else {
							deps.push(new Element(
								history.version,
								history.icon + ' ' + history.operation + ' ' + history.type + ' ' + Glyph.Bullet + ' ' + history.username + ' ' +
								history.time +
								((history.comment) ? '  ' + Glyph.TripleLine + history.comment : ' '),
								(history.type === 'branch') ? TreeItemType.View : TreeItemType.Element, history.data, vscode.TreeItemCollapsibleState.None));
						}
					});
					break;
				default: case vscode.TreeItemCollapsibleState.None:
					break;
			}
		} else {
			if (this.historyData.length > 0) {
				deps.push(new Element("Changes", ' ', TreeItemType.Root, this.historyData, vscode.TreeItemCollapsibleState.Expanded));
			} else {
				deps.push(new Element("Changes", ' Click ' + Glyph.DownwardsArrow + ' to fetch history of current file', TreeItemType.Root, this.historyData, vscode.TreeItemCollapsibleState.None));
			}
		}
		return Promise.resolve(deps);
	}
}


export class Element extends vscode.TreeItem {
	constructor(
		public readonly label: string,
		private version: string,
		private treeItemType: TreeItemType,
		public readonly historyData: HistoryData[],
		public readonly collapsibleState: vscode.TreeItemCollapsibleState,
		public readonly command?: vscode.Command
	) {
		super(label, collapsibleState);
		switch (treeItemType) {
			case TreeItemType.Root:
				this.iconPath = {
					light: path.join(__filename, '..', '..', 'images', 'light', 'icon-repo.svg'),
					dark: path.join(__filename, '..', '..', 'images', 'dark', 'icon-repo.svg')
				};
				this.contextValue = 'element.root';
				break;
			case TreeItemType.View:
				this.iconPath = {
					light: path.join(__filename, '..', '..', 'images', 'light', 'icon-tag.svg'),
					dark: path.join(__filename, '..', '..', 'images', 'dark', 'icon-tag.svg')
				};
				this.contextValue = 'element.view';
				break;
			case TreeItemType.Element:
				this.iconPath = {
					light: path.join(__filename, '..', '..', 'images', 'light', 'icon-element.svg'),
					dark: path.join(__filename, '..', '..', 'images', 'dark', 'icon-element.svg')
				};
				this.contextValue = 'element.element';
				break;
			default:
				this.iconPath = {
					light: path.join(__filename, '..', '..', 'images', 'light', 'icon-repo.svg'),
					dark: path.join(__filename, '..', '..', 'images', 'dark', 'icon-repo.svg')
				};
				this.contextValue = 'element.none';
				break;
		}
	}

	get tooltip(): string {
		return `${this.label}-${this.version}`;
	}

	get description(): string {
		return this.version;
	}
}

src/cleartool.ts
'use strict';
import * as vscode from 'vscode';
import { exec, spawn, ChildProcess, ExecException } from 'child_process';
import { HistoryProvider } from './historyProvider';
import * as ui from './userinterface';
import { LogCat } from './logcat';
import { Glyph, fetchRealLocation, datePriorToNow } from './common';
import { HistoryData } from './historyProvider';
import { join } from 'path';

export enum CleartoolCommand {
	CheckIn = 'ci',
	CheckOut = 'co',
	UndoCheckOut = 'unco',
	MakeElement = 'mkelem',
	Describe = 'describe',
	ListHistory = 'lshist',
	Differantiate = 'diff'
}



export class ClearTool {
	private tool: string;
	private historyProvider: HistoryProvider;
	private disposables: vscode.Disposable[];

	constructor() {
		this.tool = "cleartool";
		this.historyProvider = new HistoryProvider();
		this.disposables = [];
		ui.iniStatusBar();
		this.initCleartool();
	}

	createResourceUri(relativePath: string): vscode.Uri {
		let absolutePath;
		if(vscode.workspace.rootPath){
			absolutePath = join(vscode.workspace.rootPath, relativePath);
		}else{
			absolutePath = '/';
		}
		return vscode.Uri.file(absolutePath);
	}

	private initCleartool() {


		let clearcase: vscode.SourceControl = vscode.scm.createSourceControl('ClearCase', 'Clearcase');
		let checkoutgroup: vscode.SourceControlResourceGroup = clearcase.createResourceGroup('ClearCase', 'Checked-Out');
		let groupData:vscode.SourceControlResourceState[] = checkoutgroup.resourceStates;
		/*
		checkoutgroup.resourceStates = [
			{ resourceUri: this.createResourceUri('readme.md') },
			{ resourceUri: this.createResourceUri('salben32i.md') }
		];*/

		//lsco
		//(?<=checkout\sversion\s"\.\\).*(?=")

		this.disposables.push(vscode.commands.registerCommand('extension.describe', () => {
			ui.progressStatusBar(ui.StatusBarItemType.FileState, 'Initialise');
			this.describeFile(vscode.window.activeTextEditor);
		}));

		this.disposables.push(vscode.commands.registerCommand('extension.makeelement', (uri: vscode.Uri) => {
			ui.progressStatusBar(ui.StatusBarItemType.FileState, 'Creating Element...', (progress,finish) => {
				progress.report({ increment: 10, message: "Fetching file..." });
				fetchRealLocation(uri.fsPath)
					.then((realFilePath: string) => {
						progress.report({ increment: 10, message: "Check configuration..." });
						ui.showCommentDialog((param: string) => {
							progress.report({ increment: 20, message: "Execute..." });
							this.runCommand(CleartoolCommand.MakeElement, param, realFilePath)
								.then((resolve: string) => {
									progress.report({ increment: 30, message: "Finishing..." });
									ui.showMessage(resolve, ui.NotificationType.Information);
									this.describeRealFile(realFilePath , progress , finish);
								})
								.catch((reject: string) => {
									progress.report({ increment: 60, message: "Finishing..." });
									ui.showMessage(reject, ui.NotificationType.Error);
								});
						});
					}).catch(() => {
						ui.hideStatusBar();
						finish();
					});
			});
		}));

		this.disposables.push(vscode.commands.registerCommand('extension.undocheckout', (uri: vscode.Uri) => {
			ui.progressStatusBar(ui.StatusBarItemType.FileState, 'Undo Check Out', (progress,finish) => {
				progress.report({ increment: 10, message: "Fetching file..." });
				fetchRealLocation(uri.fsPath)
					.then((realFilePath: string) => {
						progress.report({ increment: 10, message: "Check configuration..." });
						ui.showUncoQuickPick((param: string) => {
							progress.report({ increment: 20, message: "Execute..." });
							this.runCommand(CleartoolCommand.UndoCheckOut, param, realFilePath)
								.then((resolve: string) => {
									progress.report({ increment: 30, message: "Finishing..." });
									ui.showMessage(resolve, ui.NotificationType.Information);
									this.describeRealFile(realFilePath , progress , finish);
								})
								.catch((reject: string) => {
									progress.report({ increment: 60, message: "Finishing..." });
									ui.showMessage(reject, ui.NotificationType.Error);
									finish();
								});
						});
					}).catch(() => {
						ui.hideStatusBar();
						finish();
					});
			});
		}));

		this.disposables.push(vscode.commands.registerCommand('extension.checkout', (uri: vscode.Uri) => {
			ui.progressStatusBar(ui.StatusBarItemType.FileState, 'Checking Out', (progress,finish) => {
				progress.report({ increment: 10, message: "Fetching file..." });
				fetchRealLocation(uri.fsPath)
					.then((realFilePath: string) => {
						progress.report({ increment: 10, message: "Check configuration..." });
						ui.showCommentDialog((param: string) => {
							progress.report({ increment: 20, message: "Execute..." });
							this.runCommand(CleartoolCommand.CheckOut, param, realFilePath)
								.then((resolve: string) => {
									progress.report({ increment: 30, message: "Finishing..." });
									ui.showMessage(resolve, ui.NotificationType.Information);
									this.describeRealFile(realFilePath , progress , finish);
								})
								.catch((reject: string) => {
									progress.report({ increment: 60, message: "Finishing..." });
									ui.showMessage(reject, ui.NotificationType.Error);
								});
						});
					}).catch(() => {
						ui.hideStatusBar();
						finish();
					});
			});
		}));

		this.disposables.push(vscode.commands.registerCommand('extension.checkin', (uri: vscode.Uri) => {
			ui.progressStatusBar(ui.StatusBarItemType.FileState, 'Checking In', (progress,finish) => {
				progress.report({ increment: 10, message: "Fetching file..." });
				fetchRealLocation(uri.fsPath)
					.then((realFilePath: string) => {
						ui.showCommentDialog((param: string) => {
							progress.report({ increment: 10, message: "Check configuration..." });
							let ptime: string = (vscode.workspace.getConfiguration("cleartool").get("actionPreserveFileModificationTime")) ? ' –ptime' : '';
							progress.report({ increment: 20, message: "Execute..." });
							this.runCommand(CleartoolCommand.CheckIn, param + ptime, realFilePath)
								.then((resolve: string) => {
									progress.report({ increment: 30, message: "Finishing..." });
									ui.showMessage(resolve, ui.NotificationType.Information);
									this.describeRealFile(realFilePath , progress , finish);
								})
								.catch((reject: string) => {
									progress.report({ increment: 60, message: "Finishing..." });
									ui.showMessage(reject, ui.NotificationType.Error);
								});
						});
				}).catch(() => {
					ui.hideStatusBar();
					finish();
				});
			});
		}));

		this.disposables.push(vscode.commands.registerCommand('cleartool.compareWithPredecessor', () => {
			if (vscode.window.activeTextEditor && vscode.window.activeTextEditor.document.uri.toString().startsWith("file:")) {
				fetchRealLocation(vscode.window.activeTextEditor.document.fileName)
					.then((realFilePath: string) => {
						this.compareWithPredecessor(realFilePath, this.historyProvider);
					}).catch(() => ui.hideStatusBar());
			}
		}));

		//todo: clear current history immediatelly : further we cache those
		this.disposables.push(vscode.commands.registerCommand('cleartool.showHistory', () => {
			ui.progressStatusBar(ui.StatusBarItemType.None, 'Show History', (progress, finish) => {
				progress.report({ increment: 10, message: "Initialise History..." });
				if (vscode.window.activeTextEditor && vscode.window.activeTextEditor.document.uri.toString().startsWith("file:")) {
					progress.report({ increment: 10, message: "Fetching file..." });
					fetchRealLocation(vscode.window.activeTextEditor.document.fileName)
						.then((realFilePath: string) => {
							progress.report({ increment: 20, message: "Parsing history..." });
							this.parseHistory(realFilePath, this.historyProvider);
							progress.report({ increment: 60, message: "Finishing..." });
							finish();
						}).catch(() => {
							ui.hideStatusBar();
							finish();
						});
				} else {
					ui.setContextCriteria(ui.FileState.Unknown);
					ui.hideStatusBar();
					finish();
				}
			});
		}));

		this.disposables.push(vscode.commands.registerCommand('cleartool.refreshHistory', () => this.historyProvider.refresh()));

		this.disposables.push(vscode.window.registerTreeDataProvider('elementHistory', this.historyProvider));

		this.disposables.push(vscode.window.onDidChangeActiveTextEditor((textEditor: vscode.TextEditor | undefined): void => {
			if (textEditor && textEditor.document.uri.toString().startsWith("file:")) {

				checkoutgroup.resourceStates.push({resourceUri:textEditor.document.uri});


				LogCat.getInstance().log('Active Editor Changed');
				ui.progressStatusBar(ui.StatusBarItemType.FileState, 'Describe');
				fetchRealLocation(textEditor.document.fileName)
					.then((realFilePath: string) => {
						this.describeRealFile(realFilePath);
					}).catch(() => {
						ui.hideStatusBar();
					});
			} else {
				ui.setContextCriteria(ui.FileState.Unknown);
				ui.hideStatusBar();
			}
		}));
		LogCat.getInstance().log('Aramok\'s Clearcase extension initialised');
	}














	
	private runCommand(command: CleartoolCommand, param: string | null, path: string): Promise<string> {
		return new Promise((resolve, reject) => {
			LogCat.getInstance().log('execute : "' + this.tool + ' ' + command + ((param) ? ' ' + param + ' ' : ' ') + path + '"');
			exec(this.tool + ' ' + command + ((param) ? ' ' + param + ' ' : ' ') + path, (error, stdout, stderr) => {
				if (error) {
					reject(error);
				} else {
					if (stderr) {
						reject(stderr);
					} else {
						resolve(stdout);
					}
				}
			});
		});
	}

	private compareWithPredecessor(realFilePath: string, historyProvider: HistoryProvider) {
		this.runCommand(CleartoolCommand.Differantiate, null, realFilePath)
			.then((resolve: string) => {
				let clearcase: vscode.SourceControl = vscode.scm.createSourceControl('ClearCase', 'Clearcase');
				let rg1: vscode.SourceControlResourceGroup = clearcase.createResourceGroup('ClearCase', 'CheckOuts (Reserved)');
				let rg2: vscode.SourceControlResourceGroup = clearcase.createResourceGroup('ClearCase', 'CheckOuts (Unreserved)');
				if (vscode.window.activeTextEditor) {
					rg1.resourceStates.push(
						{
							resourceUri: vscode.window.activeTextEditor.document.uri,
							command: undefined
						}
					);
				}
			})
			.catch((reject: string) => {
				ui.showMessage(reject, ui.NotificationType.Error);
			});
	}

	private describeFile(textEditor: vscode.TextEditor | undefined) {
		if (textEditor && textEditor.document.uri.toString().startsWith("file:")) {
			fetchRealLocation(textEditor.document.fileName)
				.then((realFilePath: string) => {
					this.describeRealFile(realFilePath);
				});
		} else {
			ui.setContextCriteria(ui.FileState.Unknown);
			ui.hideStatusBar();
		}
	}

	private describeRealFile(realFilePath: string ,
		progress?: vscode.Progress<{ message?: string | undefined; increment?: number | undefined; }>, finish?: () => void) {
		let matches: RegExpMatchArray | null;
		this.runCommand(CleartoolCommand.Describe, null, realFilePath)
			.then((resolve: string) => {
				if(progress){progress.report({message:'Command executed' , increment: 10});}
				//todo:fix regexp mathes with 'CHECKEDOUT' identifier
				let result: String = new String(resolve);
				if ((matches = result.match(/(?<=@@\\main\\).*(?=")/)) !== null) {
					if(progress){progress.report({message:'Command executed' , increment: 10});}
					let user: RegExpMatchArray | null = result.match(/(?<=\()\S*(?=\.)/);
					if (result.indexOf("CHECKEDOUT") !== -1) {
						ui.setContextCriteria(ui.FileState.CheckedOut);
						ui.viewStatus.text = `$(git-branch) ${matches.toString()}`;
						ui.fileStatus.text = `$(history) ` + ((user) ? user.toString() : ' ') + datePriorToNow(result.match(/(?<=checked out\s)(\S)*/));
						ui.fileState.text = `$(verified) Checked Out`;
					} else {
						ui.setContextCriteria(ui.FileState.Locked);
						ui.viewStatus.text = `$(git-branch) ${matches.toString()}`;
						ui.fileStatus.text = `$(history) ` + ((user) ? user.toString() : ' ') + datePriorToNow(result.match(/(?<=created\s)(\S)*(?=,|\+)/));
						ui.fileState.text = `$(lock) Locked`;
					}
				} else {
					ui.setContextCriteria(ui.FileState.Private);
					ui.viewStatus.text = `$(git-branch) No Version`;
					ui.fileStatus.text = `$(history) ` + ' ' + Glyph.EarthGround + ' -- ' + datePriorToNow(result.match(/(?<=Modified:\s).*/));
					ui.fileState.text = `$(file-code) Private File`;
				}
				if(progress){progress.report({message:'Showing Result...' , increment: 10});}
				if(finish){finish();}
			})
			.catch((reject: string) => {
				ui.setContextCriteria(ui.FileState.Unknown);
				ui.fileState.text = `$(issue-reopened) Error`;
				ui.fileStatus.hide();
				ui.viewStatus.text = `$(git-branch) No Version`;
				if(progress){progress.report({message:'Showing Result...' , increment: 30});}
				if(finish){finish();}
			});
	}

	private parseHistory(realFilePath: string, historyProvider: HistoryProvider) {
		let datas: HistoryData[] = [];
		this.runCommand(CleartoolCommand.ListHistory, null, realFilePath)
			.then((resolve: string) => {
				let result: String = new String(resolve);
				var lines = result.split('\n');
				let lastData: HistoryData;
				lines.forEach(line => {
					if (!line.startsWith('  ')) {
						//(?<=@@\\|from\s)\S*(?="|\s) // match /main/
						//(?<=@@\\)\S*(?=")
						//(?<=@@\\|from\s\\)\S*(?=\s|")
						let root: RegExpMatchArray | null = line.match(/(?<=@@\\|from\s\\)[\w|\\]*/);
						let matches: RegExpMatchArray | null = line.split(/\s+/);
						if (root) {
							if (matches) {
								let time: string = matches[0].toString();
								let username: string = matches[1].toString();
								let operation: string = matches[2].toString();
								let type: string = matches[3].toString();
								let delimiter: string[] = root.toString().split("\\");
								let seeker: HistoryData[] = datas;
								delimiter.forEach((version, index) => {
									let dataFound: HistoryData | undefined;
									dataFound = seeker.find(data => data.version === version);
									if (dataFound === undefined) {
										if (operation === 'checkout') {
											seeker.push(lastData = {
												version: version,
												icon: Glyph.CheckMark,
												username: username,
												operation: operation,
												type: type,
												time: time,
												data: []
											});
											seeker = lastData.data;
										} else {
											seeker.push(lastData = {
												version: version,
												icon: Glyph.ReturningArrow,
												username: username,
												operation: operation,
												type: type,
												time: time,
												data: []
											});
											seeker = lastData.data;
										}
									} else {
										if (index === delimiter.length - 1) {
											switch (type) {
												case 'version':
													if (dataFound.operation === 'checkout') {
														seeker.push(lastData = {
															version: version,
															icon: Glyph.ReturningArrow,
															username: username,
															operation: operation,
															type: type,
															time: time,
															data: []
														});
													}
													break;
												case 'branch':
													dataFound.data.push(lastData = {
														version: version,
														icon: Glyph.PlusSign,
														username: username,
														operation: operation,
														type: type,
														time: time,
														data: []
													});
													break;
											}
										}
										seeker = dataFound.data;
									}
								});
								if (seeker && type === 'branch') {
									//seeker.push({ version: delimiter[delimiter.length - 1] , username:username , operation:operation, type:type , time:time, data: [] });
								}
							}
						} else {
							let fname: RegExpMatchArray | null = line.match(/(?<="\.\\).*(?=@@")/);
							if (fname) {
								//todo: fix
								let time: string = matches[0].toString();
								let username: string = matches[1].toString();
								let operation: string = matches[2].toString() + ' ' + matches[3].toString();
								let type: string = matches[4].toString();
								datas.push(lastData = {
									version: fname.toString(),
									icon: Glyph.PlusSign,
									username: username,
									operation: operation,
									type: type,
									time: time,
									data: []
								});
							}
						}
					} else {
						lastData.comment = line;
					}
				});
				historyProvider.fetchHistory(datas);
			})
			.catch((reject: string) => {
				ui.showMessage(reject, ui.NotificationType.Error);
			});
	}


	public getDisposables(): vscode.Disposable[] {
		return this.disposables;
	}
}
